File: client/src/utils/verseUtils.ts
/**
 * Extract verse text from chapter content by verse number
 * Chapter content format: "1 Text\n2 Text\n3 Text..."
 * 
 * @param content - The full chapter content
 * @param verseNumber - The verse number to extract
 * @returns The verse text (including verse number) or null if not found
 */
export function getVerseText(content: string, verseNumber: number): string | null {
  if (!content) return null;
  
  const lines = content.split('\n');
  
  for (const line of lines) {
    // Match lines starting with verse number: "4 Text..." or "4\tText..."
    const match = line.match(/^(\d+)\s+(.+)$/);
    if (match) {
      const verseNum = parseInt(match[1], 10);
      const verseText = match[2].trim();
      
      if (verseNum === verseNumber) {
        return `${verseNumber} ${verseText}`;
      }
    }
  }
  
  return null;
}

/**
 * Parse comment text for verse references using #number syntax
 * 
 * @param text - The comment text to parse
 * @returns Array of verse reference objects with verse number and positions
 */
export function parseVerseReferences(text: string): Array<{verse: number, start: number, end: number}> {
  const references: Array<{verse: number, start: number, end: number}> = [];
  const regex = /#(\d+)/g;
  let match;
  
  while ((match = regex.exec(text)) !== null) {
    references.push({
      verse: parseInt(match[1], 10),
      start: match.index,
      end: match.index + match[0].length
    });
  }
  
  return references;
}

/**
 * Split comment text into segments (text and verse references)
 * 
 * @param text - The comment text
 * @returns Array of segments, each is either a text string or a verse reference object
 */
export function splitCommentContent(text: string): Array<string | {verse: number, text: string}> {
  const references = parseVerseReferences(text);
  
  if (references.length === 0) {
    return [text];
  }
  
  const segments: Array<string | {verse: number, text: string}> = [];
  let lastIndex = 0;
  
  for (const ref of references) {
    // Add text before this reference
    if (ref.start > lastIndex) {
      const textSegment = text.substring(lastIndex, ref.start);
      if (textSegment) {
        segments.push(textSegment);
      }
    }
    
    // Add the verse reference
    segments.push({
      verse: ref.verse,
      text: text.substring(ref.start, ref.end)
    });
    
    lastIndex = ref.end;
  }
  
  // Add remaining text after last reference
  if (lastIndex < text.length) {
    const textSegment = text.substring(lastIndex);
    if (textSegment) {
      segments.push(textSegment);
    }
  }
  
  return segments;
}

File: ios/ChapterADay/Utilities/VerseUtils.swift
//
//  VerseUtils.swift
//  ChapterADay
//
//  Created on 2026-01-27.
//

import Foundation

enum CommentSegment {
    case text(String)
    case verseReference(verseNumber: Int, text: String)
}

enum VerseUtils {

    /// Extract verse text from chapter content by verse number.
    /// Chapter content format: "1 Text\n2 Text\n3 Text..."
    static func getVerseText(content: String, verseNumber: Int) -> String? {
        guard !content.isEmpty else { return nil }

        let lines = content.components(separatedBy: .newlines)
        let pattern = #"^(\d+)\s+(.+)$"#
        guard let regex = try? NSRegularExpression(pattern: pattern) else { return nil }

        for line in lines {
            let range = NSRange(line.startIndex..., in: line)
            guard let match = regex.firstMatch(in: line, range: range),
                  let verseNumRange = Range(match.range(at: 1), in: line),
                  let verseTextRange = Range(match.range(at: 2), in: line) else { continue }

            let verseNum = Int(line[verseNumRange]) ?? 0
            let verseText = String(line[verseTextRange]).trimmingCharacters(in: .whitespaces)

            if verseNum == verseNumber {
                return "\(verseNumber) \(verseText)"
            }
        }
        return nil
    }

    /// Parse comment text for verse references using #number syntax.
    static func parseVerseReferences(text: String) -> [(verse: Int, start: Int, end: Int)] {
        var references: [(verse: Int, start: Int, end: Int)] = []
        let pattern = #"#(\d+)"#
        guard let regex = try? NSRegularExpression(pattern: pattern) else { return [] }

        let range = NSRange(text.startIndex..., in: text)
        regex.enumerateMatches(in: text, range: range) { match, _, _ in
            guard let match,
                  let verseRange = Range(match.range(at: 1), in: text),
                  let fullRange = Range(match.range, in: text) else { return }
            let verseNum = Int(text[verseRange]) ?? 0
            references.append((verse: verseNum, start: text.distance(from: text.startIndex, to: fullRange.lowerBound), end: text.distance(from: text.startIndex, to: fullRange.upperBound)))
        }
        return references
    }

    /// Split comment text into segments (text and verse references).
    static func splitCommentContent(text: String) -> [CommentSegment] {
        let references = parseVerseReferences(text: text)
        guard !references.isEmpty else { return [.text(text)] }

        var segments: [CommentSegment] = []
        var lastIndex = text.startIndex

        for ref in references {
            let startIdx = text.index(text.startIndex, offsetBy: ref.start)
            let endIdx = text.index(text.startIndex, offsetBy: ref.end)

            if startIdx > lastIndex {
                let textSegment = String(text[lastIndex..<startIdx])
                if !textSegment.isEmpty {
                    segments.append(.text(textSegment))
                }
            }

            segments.append(.verseReference(verseNumber: ref.verse, text: String(text[startIdx..<endIdx])))
            lastIndex = endIdx
        }

        if lastIndex < text.endIndex {
            let textSegment = String(text[lastIndex...])
            if !textSegment.isEmpty {
                segments.append(.text(textSegment))
            }
        }

        return segments
    }
}
